src/tracing/marching_squares.cpp

```diff
@@ -141,23 +152,121 @@ static void marching_squares(bool topLeft, bool topRight, bool bottomRight, bool
     }
 }
 
+/*
+ * Find other vertex with same position.
+ * @param this_index Vertex index of vertex to find other vertex with same position.
+ * @param vertices The vertices.
+ * @param out_found_index Variable to store vertex index of the found vertex.
+ * @todo Optimize using lookup table.
+ */
+static bool find_equal_vertex(const std::size_t this_index, const std::vector<vertex>& vertices, std::size_t& out_found_index) {
+    const vertex& this_vertex = vertices[this_index];
+    for (std::size_t vertex_index = 0; vertex_index < vertices.size(); ++vertex_index) {
+        const vertex& vertex = vertices[vertex_index];
+        if (this_index != vertex_index && vertex.position.x == this_vertex.position.x && this_vertex.position.y == vertex.position.y) {
+            out_found_index = vertex_index;
+            return true;
+        }
+    }
+    return false;
+}
+
+/*
+ * Tracing along contour to find all vertices of an outline.
+ * @param vertex_index Vertex index of vertex to find other vertex with same position.
+ * @param vertices The vertices generated from marching squares.
+ * @param lines The lines generated from marching squares.
+ * @param out_contour Vector to store vertices.
+ */
+static void trace_contour(const std::size_t vertex_index, std::vector<vertex>& vertices, const std::vector<line>& lines, std::vector<std::size_t>& out_contour) {
+    // Add vertex to contour.
+    out_contour.push_back(vertex_index);
+
+    // Find connected vertex on line.
+    const line& line = lines[vertices[vertex_index].line_index];
+    std::size_t connected_vertex_index = line.vertex_indices[0] == vertex_index ? line.vertex_indices[1] : line.vertex_indices[0];
```

Comment:
Use parantheses to clarify if this means:
a == (b ? t : f)
or
(a == b) ? t : f

Not strictly necessary but will make code more intuitive, i.m.o.